src/parser.rs

```diff
@@ -59,106 +104,528 @@ impl Header {
     }
 }
 
-/// Parses the different kinds of SomeIP messages (at the moment only basic type) as Result.
-#[deprecated(note = "use Someip::from_slice instead")]
-pub fn someip_message(input: &[u8]) -> Result<Message, Error> {
-    Message::from_slice(input)
-}
+impl SdPayload {
+    pub fn from_slice(s: &[u8]) -> Result<SdPayload, Error> {
+        if s.len() < SD_HEADER_LENGTH {
+            return Err(Error::NotEnoughData {
+                min: SD_HEADER_LENGTH,
+                actual: s.len(),
+            });
+        }
+        let mut buf = Cursor::new(s);
+        SdPayload::from_reader(&mut buf)
+    }
 
-impl<'a> Message<'a> {
-    /// Parse a SomeIP message from a byte array
-    pub fn from_slice(input: &'a [u8]) -> Result<Message<'a>, Error> {
-        match Header::from_slice(input) {
-            Ok(Header {
-                message_id:
-                    MessageId {
-                        service_id: 0xFFFF,
-                        method_id: 0x0000,
-                    },
-                length: 8,
-                request_id:
-                    RequestId {
-                        client_id: 0xDEAD,
-                        session_id: 0xBEEF,
-                    },
-                protocol_version: 0x01,
-                interface_version: 0x01,
-                message_type: MessageType::RequestNoReturn,
-                return_code: ReturnCode::Ok,
-            }) => Ok(Message::MagicCookieClient),
+    pub fn from_reader<R: Read>(reader: &mut R) -> Result<SdPayload, Error> {
+        let flags = reader.read_u8()?;
+        reader.read_u24::<BE>()?; // reserved
 
-            Ok(Header {
-                message_id:
-                    MessageId {
-                        service_id: 0xFFFF,
-                        method_id: 0x8000,
-                    },
-                length: 8,
-                request_id:
-                    RequestId {
-                        client_id: 0xDEAD,
-                        session_id: 0xBEEF,
-                    },
-                protocol_version: 0x01,
-                interface_version: 0x01,
-                message_type: MessageType::Notification,
-                return_code: ReturnCode::Ok,
-            }) => Ok(Message::MagicCookieServer),
+        let mut entries: Vec<SdEntry> = Vec::new();
```

Comment:
With the number of entries known the vec could be initialised with a dedicated capacity. This safes space and in best case allocations.

Replace with `Vec::with_capacity`